### PR TITLE
feat: v4 final

### PR DESCRIPTION
**Pull request by dekz**
## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts [hunk 1]
```diff
@@ -205,56 +164,68 @@ export class MarketOperationUtils {
             : Promise.resolve([]);
 
         const [
-            [tokenDecimals, orderFillableAmounts, ethToMakerAssetRate, ethToTakerAssetRate, dexQuotes, rawTwoHopQuotes],
-            rfqtIndicativeQuotes,
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/index.ts:209]: Are indicative quotes always fetched after sampling/optimization now?

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts [hunk 2]
```diff
@@ -639,34 +599,39 @@ export class MarketOperationUtils {
...
+                    takerToken,
                     amount,
-                    marketSideLiquidity,
-                    _opts.feeSchedule,
-                ).wholeOrder;
-            }
-
-            const { isFirmPriceAwareEnabled, isIndicativePriceAwareEnabled } = getPriceAwareRFQRolloutFlags(
-                rfqt.priceAwareRFQFlag,
-            );
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/index.ts:658]: Do all MMs support price-aware queries now? Looks like there's no more checking for flags.

**Review by dorothy-zbornak — APPROVED** [review]
Congrats on this, guys! I know it was a big effort.

File: packages/asset-swapper/src/utils/quote_requestor.ts [hunk 3]
```diff
@@ -111,15 +97,14 @@ export class QuoteRequestor {
 
         const requestParamsWithBigNumbers: Pick<
             TakerRequestQueryParams,
-            'buyTokenAddress' | 'sellTokenAddress' | 'takerAddress' | 'comparisonPrice' | 'protocolVersion'
+            'buyTokenAddress' | 'sellTokenAddress' | 'txOrigin' | 'comparisonPrice' | 'protocolVersion' | 'takerAddress'
         > = {
+            txOrigin,
             takerAddress,
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/utils/quote_requestor.ts:103]: yeah `takerAddress` is NULL, I was debating on where to really have that default, at the moment we just took the hit in 0xAPI.

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts (hunk 1 above)
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/index.ts:209]: Yeah so the RFQ maker can get the price to beat. Same flow for firm/indicative. Discover price to beat, RFQ, re-optimize

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts (hunk 2 above)
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/index.ts:658]: Yeah it's now the default for everyone to receive a price to beat

File: packages/asset-swapper/src/utils/swap_quote_consumer_utils.ts [hunk 4]
```diff
@@ -59,29 +56,22 @@ export const swapQuoteConsumerUtils = {
...
-    },
-    isValidForwarderSignedOrders(orders: SignedOrder[], wethAssetData: string): boolean {
-        return _.every(orders, order => swapQuoteConsumerUtils.isValidForwarderSignedOrder(order, wethAssetData));
-    },
-    isValidForwarderSignedOrder(order: SignedOrder, wethAssetData: string): boolean {
-        return isExactAssetData(order.takerAssetData, wethAssetData);
+    isValidForwarderSwapQuote(swapQuote: SwapQuote, wethToken: string): boolean {
+        return swapQuote.orders.find(o => o.takerToken !== wethToken) === undefined;
     },
     async getExtensionContractTypeForSwapQuoteAsync(
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/utils/swap_quote_consumer_utils.ts:62]: DEEEELETTEEEDDD

File: packages/asset-swapper/src/types.ts [hunk 5]
```diff
@@ -22,34 +29,24 @@ export interface OrderPrunerOpts {
...
- * makerAssetData: The assetData representing the desired makerAsset.
- * takerAssetData: The assetData representing the desired takerAsset.
- */
-export interface OrderProviderRequest {
-    makerAssetData: string;
-    takerAssetData: string;
-}
+export type SignedNativeOrder = SignedOrder<LimitOrderFields> | SignedOrder<RfqOrderFields>;
+export type NativeOrderWithFillableAmounts = SignedNativeOrder & NativeOrderFillableAmountFields;
+export type OrderWithFillableAmounts = CommonOrderFields & NativeOrderFillableAmountFields;
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/types.ts:40]: BE GONE

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts [hunk 6]
```diff
@@ -158,11 +127,14 @@ export class MarketOperationUtils {
             ...(!sampleBalancerOnChain ? [ERC20BridgeSource.Balancer] : []),
         ];
 
+        // Used to determine whether the tx origin is an EOA or a contract
+        const txOrigin = (_opts.rfqt && _opts.rfqt.txOrigin) || NULL_ADDRESS;
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/index.ts:131]: :trollface:

File: packages/asset-swapper/src/utils/market_operation_utils/orders.ts [hunk 7]
```diff
@@ -154,209 +68,158 @@ export function createOrdersFromTwoHopSample(
...
-    makerToken: string,
-    takerToken: string,
-    opts: CreateOrderFromPathOpts,
-): OptimizedMarketOrder {
-    const bridgeAddress = getBridgeAddressFromFill(fill, opts);
+export function createBridgeDataForBridgeOrder(order: OptimizedMarketBridgeOrder): string {
+    let bridgeData: string;
+    if (
+        order.source === ERC20BridgeSource.MultiHop ||
+        order.source === ERC20BridgeSource.MultiBridge ||
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/orders.ts:130]: Yeah I reckon. I'll add a TODO, need to test what happens to integrations that might exclude it or anything similar.

File: packages/asset-swapper/src/utils/market_operation_utils/types.ts [hunk 8]
```diff
@@ -153,35 +158,24 @@ export interface DODOFillData extends FillData {
...
 }
 
 /**
  * Represents a node on a fill path.
  */
-export interface Fill<TFillData extends FillData = FillData> extends SourceInfo<TFillData> {
+export interface Fill<TFillData extends FillData = FillData> {
+    // basic data for every fill
+    source: ERC20BridgeSource;
+    type: FillQuoteTransformerOrderType; // should correspond with TFillData
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/types.ts:177]: Added a TODO

File: packages/asset-swapper/src/utils/utils.ts [hunk 9]
```diff
@@ -1,134 +1,100 @@
...
+    if (takerFillAmount.isZero()) {
+        return EMPTY_FILLABLE_AMOUNTS;
+    }
+    return {
+        fillableTakerAmount: takerFillAmount,
+        fillableMakerAmount: getNativeAdjustedMakerFillAmount(order.order, takerFillAmount),
+        fillableTakerFeeAmount:
+            order.type === FillQuoteTransformerOrderType.Limit
+                ? getNativeAdjustedTakerFeeAmount(order.order as LimitOrderFields, takerFillAmount)
+                : ZERO_AMOUNT,
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/utils/utils.ts]: The input should be fillable amounts, rather then fill amount. I.e `min(total-filled, balance, allowance)`

File: packages/asset-swapper/src/swap_quoter.ts [hunk 10]
```diff
@@ -228,193 +145,56 @@ export class SwapQuoter {
...
-                return sortingUtils.sortOrders(prunedOrders);
-            }
-        });
-
-        const swapQuotes = await this._swapQuoteCalculator.calculateBatchMarketBuySwapQuoteAsync(
-            allPrunedOrders,
-            makerAssetBuyAmount,
-            gasPrice,
-            calculateSwapQuoteOpts,
+        const allOrders = await this.orderbook.getBatchOrdersAsync(
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/swap_quoter.ts:165]: ty ty we do

File: packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 11]
```diff
@@ -164,52 +179,44 @@ export class ExchangeProxySwapQuoteConsumer implements SwapQuoteConsumerBase {
             });
         }
 
-        const intermediateToken = quote.isTwoHop ? getTokenFromAssetData(quote.orders[0].makerAssetData) : NULL_ADDRESS;
+        const intermediateToken = quote.isTwoHop ? quote.orders[0].makerToken : NULL_ADDRESS;
         // This transformer will fill the quote.
         if (quote.isTwoHop) {
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts:184]: We use it down later for pay taker transformer, incase there's dust intermediate.

**Pull request merged**